plugins/cpp_metrics/parser/src/cppmetricsparser.cpp

```diff
@@ -0,0 +1,151 @@
+#include <cppmetricsparser/cppmetricsparser.h>
+
+#include <model/cppastnodemetrics.h>
+#include <model/cppastnodemetrics-odb.hxx>
+#include <model/cppfilemetrics.h>
+#include <model/cppfilemetrics-odb.hxx>
+
+#include <model/cppastnode.h>
+#include <model/cppastnode-odb.hxx>
+
+#include <boost/filesystem.hpp>
+
+#include <util/logutil.h>
+#include <util/odbtransaction.h>
+
+#include <memory>
+
+namespace cc
+{
+namespace parser
+{
+
+CppMetricsParser::CppMetricsParser(ParserContext& ctx_): AbstractParser(ctx_)
+{
+  util::OdbTransaction {_ctx.db} ([&, this] {
+    for (const model::CppFileMetrics& fm
+      : _ctx.db->query<model::CppFileMetrics>())
+    {
+      _fileIdCache.insert(fm.file);
+    }
+
+    for (const model::CppAstNodeMetrics& anm
+      : _ctx.db->query<model::CppAstNodeMetrics>())
+    {
+      auto node = _ctx.db->query_one<model::CppAstNode>(
+        odb::query<model::CppAstNode>::id == anm.astNodeId);
+      _astNodeIdCache.insert({anm.astNodeId, node->location.file->id});
+    }
+  });
+}
+
+bool CppMetricsParser::cleanupDatabase()
+{
+  if (!_fileIdCache.empty())
+  {
+    try
+    {
+      util::OdbTransaction {_ctx.db} ([this] {
+        for (const model::File& file
+          : _ctx.db->query<model::File>(
+          odb::query<model::File>::id.in_range(_fileIdCache.begin(), _fileIdCache.end())))
+        {
+          auto it = _ctx.fileStatus.find(file.path);
+          if (it != _ctx.fileStatus.end() &&
+              (it->second == cc::parser::IncrementalStatus::DELETED ||
+               it->second == cc::parser::IncrementalStatus::MODIFIED ||
+               it->second == cc::parser::IncrementalStatus::ACTION_CHANGED))
+          {
+            LOG(info) << "[cxxmetricsparser] Database cleanup: " << file.path;
+
+            _ctx.db->erase_query<model::CppFileMetrics>(odb::query<model::CppFileMetrics>::file == file.id);
+            _fileIdCache.erase(file.id);
+          }
+        }
+
+        for (const auto& pair : _astNodeIdCache)
+        {
+          auto file = _ctx.db->query_one<model::File>(
+            odb::query<model::File>::id == pair.second);
+
+          auto it = _ctx.fileStatus.find(file->path);
+          if (it != _ctx.fileStatus.end() &&
+              (it->second == cc::parser::IncrementalStatus::DELETED ||
+               it->second == cc::parser::IncrementalStatus::MODIFIED ||
+               it->second == cc::parser::IncrementalStatus::ACTION_CHANGED))
+          {
+            LOG(info) << "[cxxmetricsparser] Database cleanup: " << file->path;
+
+            _ctx.db->erase_query<model::CppAstNodeMetrics>(odb::query<model::CppAstNodeMetrics>::astNodeId == pair.first);
+            _astNodeIdCache.erase(pair.first);
+          }
+        }
```

Comment:
In the loop above you are already iterating through all files and purging the ones, which has *deleted*, *modified* or *action changed* status. There is no need for another loop or for the `_astNodeIdCache` at all, we could simply purge the AST nodes there as well.

Reply:
I will accept it like this, so we can merge this, and created #670.